webdav_storage/src/WebDav.php

```diff
@@ -0,0 +1,369 @@
+<?php
+
+namespace Friendica\Addon\webdav_storage\src;
+
+use Exception;
+use Friendica\Core\Config\IConfig;
+use Friendica\Core\L10n;
+use Friendica\Model\Storage\IWritableStorage;
+use Friendica\Model\Storage\ReferenceStorageException;
+use Friendica\Model\Storage\StorageException;
+use Friendica\Network\HTTPClientOptions;
+use Friendica\Network\IHTTPClient;
+use Friendica\Util\Strings;
+use Psr\Log\LoggerInterface;
+
+/**
+ * A WebDav Backend Storage class
+ */
+class WebDav implements IWritableStorage
+{
+	const NAME = 'WebDav';
+
+	/** @var L10n */
+	private $l10n;
+
+	/** @var IConfig */
+	private $config;
+
+	/** @var string */
+	private $url;
+
+	/** @var IHTTPClient */
+	private $client;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var array */
+	private $authOptions;
+
+	public function __construct(L10n $l10n, IConfig $config, IHTTPClient $client, LoggerInterface $logger)
```

Comment:
I feel like the `$config` parameter should be an associative array with the content of `DI::config()['webdav']` rather than the whole object?

Reply:
I don't like that the caller has to know the category of the configuration. I'd prefer that the class itself is responsible for retrieving the right config .. As a alternative solution, we could create a `WebDavSettings` class which is responsible for mapping the configuration out of the array, so it's encapsulated and reusable as well.

Reply:
The caller must have a knowledge of the category of the configuration because it will be the addon itself, right? It's better the addon knows the category of the config than the WebDav connector class.

Reply:
I'm with you, but what about the `saveOptions()` method? You have to know the `IConfig` structure at least for saving it ..

I believe that we're mixing two purposes into one class:

- A class for communication with a webdav backend
- A class to configure a webdav backend

I got at this point at the past as well, but I didn't want to touch the `IStorage` interface again ^^

Reply:
I forgot about that method, and it really isn't ideal in my mind. The addon should take care of the configuration and the class of the communication.

I can touch the interface if you'd like me to. None of the other `saveOptions` implementations are hard to refactor.

Reply:
No it's fine, I just needed another similar opinion and now I'm motivated again *gg* .. Before this PR I came to the conclusion not to re-re-refactor things because of beauty issues ;-)